codegen/ace/src/encoded.rs

```diff
@@ -0,0 +1,280 @@
+    /// the following order, from least significant:
+    /// - 30 bits for the right index
+    /// - 30 bits for the left index
+    /// - 2 bits for the OPCODE, with SUB 00, MUL 01, ADD 10
+    ///
+    /// The output indexes of inputs, constants and gates are not serialized as they
+    /// are simply successive values starting from `n_inputs` to
+    /// `n_inputs + n_constants + n_gates - 1`.
+    ///
+    /// For example, the circuit `(i0 + 14) * i0` encodes to
+    /// ```text
+    /// [
+    /// 14, // value of the constant, which has index 1, given that there is one input at index 0
+    /// (2 << 60) + (0 << 30) + (1 << 0), // addition gate opcode with left input 0 and right input 1. Output index 2.
+    /// (1 << 60) + (2 << 30) + (0 << 0), // multiplication gate opcode with left input 2 and right input 0. Output index 3.
+    /// ]
+    /// ```
+    pub fn to_ace(&self) -> EncodedCircuit {
+        const MAX_NODE_ID: u64 = (1 << 30) - 1;
+        // assert!(self.is_padded(), "circuit not padded");
+
+        assert!(
+            self.num_nodes() as u64 <= MAX_NODE_ID,
+            "more than 2^30 nodes"
+        );
+
+        // Constants are encoded two-by-two as extension field elements, followed by operations.
+        let num_const = self.constants.len().next_multiple_of(2);
+        let num_ops = self.operations.len().next_multiple_of(4);
+        let len_const = num_const * 2;
+        let len_circuit = len_const + num_ops;
+        let mut instructions = Vec::with_capacity(len_circuit);
+
+        // Add constants
+        instructions.extend(
+            self.constants
+                .iter()
+                .flat_map(|c| QuadFelt::from(*c).to_base_elements()),
+        );
+        // Since constants are treated as extension field elements, we add an extra 0
+        // to ensure the section in the circuit is word-aligned
+        instructions.resize(len_const, Felt::ZERO);
+
+        let num_inputs = self.layout.num_inputs;
+        let num_constants = self.constants.len();
+        let num_nodes = num_inputs + num_constants + num_ops;
+        let node_id = |node: Node| -> u64 {
+            let input_start = num_nodes - 1;
+            let constants_start = input_start - num_inputs;
+            let ops_start = constants_start - num_constants;
+
+            match node {
+                Node::Input(idx) => input_start.checked_sub(idx),
+                Node::Constant(idx) => constants_start.checked_sub(idx),
+                Node::Operation(idx) => ops_start.checked_sub(idx),
+            }
+            .unwrap() as u64
+        };
+
+        let operation_to_instruction = |operation: &OperationNode| {
+            let id_0 = node_id(operation.node_l);
+            let id_1 = node_id(operation.node_r);
+            let op_tag = operation.op as u64;
+
+            const OP_OFFSET: u64 = 1 << 60;
+            const ID_1_OFFSET: u64 = 1 << 30;
```

Comment:
Do we need these elsewhere? If so, it might be worth it to define them outside

Reply:
Not entirely sure what to do with them. We could make them public and import them from the VM, but this could be done when we import the AirScript from the VM with the generated verifier.